File: Cindy/Health/ReadinessEstimator.swift
import Foundation

/// What Apple Health contributes to the readiness estimate. All optional:
/// without an Apple Watch most of it stays empty, and the estimate falls back
/// to the workout history alone.
struct HealthMetrics: Equatable {
    var heartRateVariability: MetricBaseline?
    var restingHeartRate: MetricBaseline?
    /// Hours actually asleep last night (overlapping sources merged).
    var sleepHours: Double?
    var bodyMassKilograms: Double?

    static let none = HealthMetrics()

    var isEmpty: Bool {
        heartRateVariability == nil && restingHeartRate == nil && sleepHours == nil
    }
}

/// Turns the workout history and the Health metrics into a `Readiness`.
/// Pure: no HealthKit, no clock of its own, so it can be unit-tested.
struct ReadinessEstimator {
    var now: Date = Date()

    func estimate(history: [WorkoutRecord], metrics: HealthMetrics = .none) -> Readiness? {
        var components: [Readiness.Component] = []

        if let last = history.max(by: { $0.date < $1.date }) {
            let hours = now.timeIntervalSince(last.date) / 3600
            let needed = ReadinessScoring.recoveryHours(for: last)
            components.append(component(.recovery,
                                        ReadinessScoring.recovery(fraction: max(hours, 0) / needed),
                                        L("\(Int(max(hours, 0).rounded())) h since the last workout, \(Int(needed.rounded())) h suggested")))
        }
        if let ratio = ReadinessEstimator.loadRatio(history: history, now: now) {
            components.append(component(.trainingLoad,
                                        ReadinessScoring.trainingLoad(ratio: ratio),
                                        L("This week's volume is \(ratio.formattedDecimal(1))× your average")))
        }
        if let hours = metrics.sleepHours {
            components.append(component(.sleep, ReadinessScoring.sleep(hours: hours),
                                        L("\(hours.formattedDecimal(1)) h of sleep last night")))
        }
        if let baseline = metrics.heartRateVariability, let z = baseline.zScore {
            components.append(component(.heartRateVariability,
                                        ReadinessScoring.heartRateVariability(zScore: z),
                                        L("HRV \(Int(baseline.value.rounded())) ms, your average \(Int(baseline.mean.rounded())) ms")))
        }
        if let baseline = metrics.restingHeartRate, let z = baseline.zScore {
            components.append(component(.restingHeartRate,
                                        ReadinessScoring.restingHeartRate(zScore: z),
                                        L("Resting pulse \(Int(baseline.value.rounded())) bpm, your average \(Int(baseline.mean.rounded())) bpm")))
        }

        return Readiness(components: components, usesHealthData: !metrics.isEmpty)
    }

    /// Volume of the last 7 days divided by the weekly average of the last 28,
    /// counted in reps. `nil` below three sessions, where the average says nothing.
    static func loadRatio(history: [WorkoutRecord], now: Date, minimumSessions: Int = 3) -> Double? {
        let week = now.addingTimeInterval(-7 * 24 * 3600)
        let month = now.addingTimeInterval(-28 * 24 * 3600)
        let recent = history.filter { $0.date > month && $0.date <= now }
        guard recent.count >= minimumSessions else { return nil }
        let chronicWeekly = Double(recent.reduce(0) { $0 + $1.score.totalReps }) / 4
        guard chronicWeekly > 0 else { return nil }
        let acute = Double(recent.filter { $0.date > week }.reduce(0) { $0 + $1.score.totalReps })
        return acute / chronicWeekly
    }

    private func component(_ kind: Readiness.Component.Kind, _ score: Double,
                           _ detail: String) -> Readiness.Component {
        Readiness.Component(kind: kind, score: score,
                            weight: ReadinessScoring.weights[kind] ?? 0, detail: detail)
    }
}

File: android/core/src/main/kotlin/me/raddatz/cindy/core/health/ReadinessEstimator.kt
package me.raddatz.cindy.core.health

import me.raddatz.cindy.core.addingSeconds
import me.raddatz.cindy.core.persistence.WorkoutRecord
import me.raddatz.cindy.core.roundedHalfAwayFromZero
import me.raddatz.cindy.core.secondsSince
import java.time.Instant

/**
 * What the health store contributes to the readiness estimate. All optional: without a watch
 * most of it stays empty, and the estimate falls back to the workout history alone.
 */
data class HealthMetrics(
    /**
     * Heart rate variability against its baseline. iOS reads SDNN, Health Connect only offers
     * RMSSD; the estimator only looks at the z-score, so either works — but never mix both in
     * one baseline.
     */
    val heartRateVariability: MetricBaseline? = null,
    val restingHeartRate: MetricBaseline? = null,
    /** Hours actually asleep last night (overlapping sources merged). */
    val sleepHours: Double? = null,
    val bodyMassKilograms: Double? = null,
) {
    val isEmpty: Boolean
        get() = heartRateVariability == null && restingHeartRate == null && sleepHours == null

    companion object {
        val none: HealthMetrics = HealthMetrics()
    }
}

/**
 * Turns the workout history and the health metrics into a [Readiness].
 * Pure: no health store, no clock of its own, so it can be unit-tested.
 */
data class ReadinessEstimator(val now: Instant = Instant.now()) {

    fun estimate(history: List<WorkoutRecord>, metrics: HealthMetrics = HealthMetrics.none): Readiness? {
        val components = mutableListOf<Readiness.Component>()

        val last = history.maxByOrNull { it.date }
        if (last != null) {
            val hours = now.secondsSince(last.date) / 3600
            val needed = ReadinessScoring.recoveryHours(last)
            components += component(
                Readiness.Component.Kind.RECOVERY,
                ReadinessScoring.recovery(fraction = maxOf(hours, 0.0) / needed),
                ReadinessDetail.SinceLastWorkout(
                    hoursSince = maxOf(hours, 0.0).roundedHalfAwayFromZero().toInt(),
                    hoursSuggested = needed.roundedHalfAwayFromZero().toInt(),
                ),
            )
        }
        val ratio = loadRatio(history, now)
        if (ratio != null) {
            components += component(
                Readiness.Component.Kind.TRAINING_LOAD,
                ReadinessScoring.trainingLoad(ratio),
                ReadinessDetail.TrainingLoad(ratio),
            )
        }
        val sleepHours = metrics.sleepHours
        if (sleepHours != null) {
            components += component(
                Readiness.Component.Kind.SLEEP,
                ReadinessScoring.sleep(sleepHours),
                ReadinessDetail.Sleep(sleepHours),
            )
        }
        val hrv = metrics.heartRateVariability
        val hrvZ = hrv?.zScore
        if (hrv != null && hrvZ != null) {
            components += component(
                Readiness.Component.Kind.HEART_RATE_VARIABILITY,
                ReadinessScoring.heartRateVariability(hrvZ),
                ReadinessDetail.HeartRateVariability(
                    hrv.value.roundedHalfAwayFromZero().toInt(),
                    hrv.mean.roundedHalfAwayFromZero().toInt(),
                ),
            )
        }
        val pulse = metrics.restingHeartRate
        val pulseZ = pulse?.zScore
        if (pulse != null && pulseZ != null) {
            components += component(
                Readiness.Component.Kind.RESTING_HEART_RATE,
                ReadinessScoring.restingHeartRate(pulseZ),
                ReadinessDetail.RestingHeartRate(
                    pulse.value.roundedHalfAwayFromZero().toInt(),
                    pulse.mean.roundedHalfAwayFromZero().toInt(),
                ),
            )
        }

        return Readiness.from(components, usesHealthData = !metrics.isEmpty)
    }

    private fun component(kind: Readiness.Component.Kind, score: Double, detail: ReadinessDetail) =
        Readiness.Component(kind, score, ReadinessScoring.weights[kind] ?: 0.0, detail)

    companion object {
        /**
         * Volume of the last 7 days divided by the weekly average of the last 28, counted in reps.
         * `null` below three sessions, where the average says nothing.
         */
        fun loadRatio(history: List<WorkoutRecord>, now: Instant, minimumSessions: Int = 3): Double? {
            val week = now.addingSeconds(-7.0 * 24 * 3600)
            val month = now.addingSeconds(-28.0 * 24 * 3600)
            val recent = history.filter { it.date > month && it.date <= now }
            if (recent.size < minimumSessions) return null
            val chronicWeekly = recent.fold(0) { sum, r -> sum + r.score.totalReps }.toDouble() / 4
            if (!(chronicWeekly > 0)) return null
            val acute = recent.filter { it.date > week }.fold(0) { sum, r -> sum + r.score.totalReps }.toDouble()
            return acute / chronicWeekly
        }
    }
}
